### PR TITLE
feat: #15416 Expand round robin availability beyond default schedule

### LINKED ISSUE #15416: [CAL-3926] Expand round robin availability beyond the default schedule
**Problem**

Currently with round robin event types, you can only use a person's default availability schedule

*User context:*

> We are reaching the limits of what we can achieve with the round robin's due to the limitations of what schedule's can be used for the [Hosts](http://hosts.ie/). They either all share a common one, which doesn't work us given the different times all our staff work, or it uses their default schedules only.
>
> We are hoping this could be expanded on to allow custom availability schedule per host for round robin event [types. As](http://types.As) we are reaching a bottle neck with our returning patients with our current use of single event types for each doctor, and the difficulty for us to check them one by one to find a doctor with a free timeslot.
>
> We are wanting to create a new round robin and enter patients back into it for follow up time slots, but without impacting the availability for new patients.

**Ideal scenario**

* Each user would be able to create a new availability schedule called “Returning patients” (already possible today)

* They could then add that specific schedule to a new round-robin event type called “Returning patient” etc. 

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3926](https://linear.app/calcom/issue/CAL-3926/expand-round-robin-availability-beyond-the-default-schedule)</sub>

### PR DESCRIPTION
**Pull request by vijayraghav-io**
## What does this PR do?

- Fixes #15416 
- Fixes CAL-3926

/claim #15416 

@ciaranha @itsberkaya 

https://www.loom.com/share/b194ce55a54649edad65eb10587994c6?sid=3faf6346-e165-4e4d-adf4-a4a9f0193bb7

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A -  I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by anikdhabal** [comment]
Hey @vijayraghav-io, great work! However, this issue needs a design, and we are waiting for the designer's approval or a new design. Make this PR a draft until we receive their input.

Thanks for your work.

**Comment by Ryukemeister** [comment]
@vijayraghav-io I know this PR is still in the draft phase, but can you have a look at why type check are failing?

**Comment by vijayraghav-io** [comment]
Yes checking

**Comment by vijayraghav-io** [comment]
> @vijayraghav-io I know this PR is still in the draft phase, but can you have a look at why type check are failing?

updated for errors, now tests are passing

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Seems to me like you forgot to include migration file, please add those changes @vijayraghav-io 🙏

File: packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts
Comment (Amit91848) [inline packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:48]: Check for user role before querying the database

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Left some comments

**Comment by vijayraghav-io** [comment]
> Seems to me like you forgot to include migration file, please add those changes @vijayraghav-io 🙏

Added migration file

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx:290]: Its just that selection is defaulted to first element in available options by setting to 0, no wrong data is sent to backend.

File: packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts
Comment (vijayraghav-io) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:48]: Here the checks  `if(!isCurrentUserPartOfTeam && !isCurrentUserOwner)` , ensures that data (i.e. availabilities) of current user's or his team member's only can be queried. It should be possible regardless of the role.

Comment (Amit91848) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:48]: I meant do this check `if(!isCurrentUserPartOfTeam && !isCurrentUserOwner)`  before querying the database for schedules

Comment (vijayraghav-io) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts]: oh, ok! done

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (Amit91848) [inline apps/web/modules/event-types/views/event-types-single-view.tsx:127]: Why are you storing AvailabilityOption? We are already fetching all schedules of the user, you can find selected Option by matching it against scheduleId.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (Amit91848) [inline apps/web/components/eventtype/EventAvailabilityTab.tsx:398]: Pass the index as props instead of having to search it using getIdx.

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (Amit91848) [inline apps/web/modules/event-types/views/event-types-single-view.tsx:129]: I don't think we should extend host to include avatar and label especially since we are not storing it. You can use getValue and find avatar and label of the host inside TeamMemberSchedule.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (Amit91848) [inline apps/web/components/eventtype/EventAvailabilityTab.tsx:338]: No need to store the availabilityValue, you can find the selected option using scheduleId and schedules.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (Amit91848) [inline apps/web/components/eventtype/EventAvailabilityTab.tsx:324]: This won't be necessary now. since we are only using scheduleId.

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Great work so far @vijayraghav-io . Left a few comments. Lmk your views.

Also when you add a host, it resets the availability of host of the same type, i.e if you add a fixed host, it changes the availability of other fixed host to default and same with round robin hosts.


https://github.com/user-attachments/assets/83935797-2988-49c9-a1a5-22f0320b1b29

**Comment by vijayraghav-io** [comment]
> Great work so far @vijayraghav-io . Left a few comments. Lmk your views.
> 
> Also when you add a host, it resets the availability of host of the same type, i.e if you add a fixed host, it changes the availability of other fixed host to default and same with round robin hosts.
> 
>  Recording.2024-07-15.160752.mp4

Thankyou @Amit91848 , i am going through the comments and will address them.

**Comment by vijayraghav-io** [comment]
> Great work so far @vijayraghav-io . Left a few comments. Lmk your views.
> 
> Also when you add a host, it resets the availability of host of the same type, i.e if you add a fixed host, it changes the availability of other fixed host to default and same with round robin hosts.
> 
>  Recording.2024-07-15.160752.mp4

I have updated fix for this and is working fine for same hosts and different hosts. My bad i was checking for different kind of hosts all the time. Great find!

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: Yeah useEffect is not required, i had started with functionality of showing availability in TeamTab that needed useEffect(for sorting on priority), its not required now, removed.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: updated

File: packages/core/getUserAvailability.ts
Comment (vijayraghav-io) [reply packages/core/getUserAvailability.ts]: i have updated to be concise, with 'undefined' tests were failing as schedule can be null, so used nullish Coalescing and verified tests are passing.

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (vijayraghav-io) [reply apps/web/modules/event-types/views/event-types-single-view.tsx:127]: This is required to remember the user selection for scenario when user is toggling between `Choose a common schedule`. When common schedule is choosen scheduleId is made null.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: was more cautious , as the hosts index is changed when priority in TeamTab (or Assignment Tab) is changed due to sorting. Since TeamMemberSchedule is child component of TeamAvailability, this is not an issue, tested and updated.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: This is required to set default values when new host is added

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (vijayraghav-io) [reply apps/web/modules/event-types/views/event-types-single-view.tsx:129]: This is only for enriching the fetched data, avatar and label are not stored through hosts, these fields are removed before sending payload.
Also wanted to be consistent with Assignment Tab. The label in EventTeamTab (Assignment) is fetched from EventSetupProps.teamMembers. This prop in not available in AvailabilityTab, else requires similar changes.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: used to set initial value, watch is not required. Updated.

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (Amit91848) [reply apps/web/modules/event-types/views/event-types-single-view.tsx:129]: My bad, I meant you can access team members from eventType that is passed as props to UseCommonScheduleSettings by eventType.team.members[number].user and get avatar and label from there. 

Extending hosts is unnecessary as then you would have to clean it up later, best use eventType which is already being passed.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (Amit91848) [inline apps/web/components/eventtype/EventAvailabilityTab.tsx:399]: This is where you can access team members from eventType.team.members for avatar and label for my previous comment

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (Amit91848) [inline apps/web/modules/event-types/views/event-types-single-view.tsx:356]: This won't be needed along with other places where you are updating host with avatar and label

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Thanks for such quick changes 🔥, left a few more suggestions @vijayraghav-io . Also once you are done please update PR description with latest changes screenshot / video, @ciaranha would need to approve the ux and design changes that were made

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (Amit91848) [inline apps/web/components/eventtype/EventAvailabilityTab.tsx:323]: You can also clean this up and use the value directly inside the Select.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: Updated

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (vijayraghav-io) [reply apps/web/modules/event-types/views/event-types-single-view.tsx]: Updated

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: scheduleId has to be initialized, else 'null' will be sent to backend when -> new host is added in Assignment Tab but user does not change the default schedule selection. As OnChange() triggers only if there is change in selection.

**Comment by vijayraghav-io** [comment]
Thanks @Amit91848 , Updated. 
Also description is updated with latest UI video.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (Amit91848) [inline apps/web/components/eventtype/EventAvailabilityTab.tsx:324]: This is still sending null to backend if you add host and save directly without making changes in the availability tab? Not sure what this is achieving? Imo this is not required. In case no scheduleId is being selected, aren't we defaulting to users default schedule anyway?


https://github.com/user-attachments/assets/99174cc4-bce3-4dc5-a1fe-7d82d862bb09

**Review by Amit91848 — CHANGES_REQUESTED** [review]

**Comment by Amit91848** [comment]
This is working for collective events too, do you want to limit it to round robin only @ciaranha ?

File: apps/web/modules/event-types/views/event-types-single-view.tsx
Comment (vijayraghav-io) [reply apps/web/modules/event-types/views/event-types-single-view.tsx:469]: The type is not defined for eventType.team.members and to avoid lint errors using any, TeamMembers under eventTypeSetupProps is used, which is used in other components as well and has the same required data.

File: apps/web/components/eventtype/EventTeamTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventTeamTab.tsx:342]: Yes this is required, else whenever a new host of same type is added , already added (or existing) host will loose its scheduleId.
So this is done, to ensure previously added hosts do no loose scheduleId.
This is because the host is created from list option in `CheckedHostField` component.

File: apps/web/components/eventtype/EventTeamTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventTeamTab.tsx:365]: Explained above

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx:324]: This may occur if host is added in Assignment Tab and form is saved without visiting Availability Tab.
Alternatively, this can be moved to submit handler, but there we will not have AvailabilityOptions to set the default schedule id. As  now we are confined AvailabilityOptions to TeamMemberSchedule component only and removed in Host. And we cant rely on .team.member.user.defaultScheduleId as this not always populated. (had checked this in trpc endpoint).

Anyhow the functionality of getUserAvailability() will still pick the default Schedule, as this how its updated at backend. So the event booking will not break.

Will update to handle this scenario, but not sure if this line can be completely removed. Will check and update.

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx:397]: I had added for <li >, will check

File: apps/web/components/eventtype/EventTeamTab.tsx
Comment (Amit91848) [reply apps/web/components/eventtype/EventTeamTab.tsx:342]: Makes sense, please add a comment because I couldn't figure out why it was added just by the code.

Comment (Amit91848) [reply apps/web/components/eventtype/EventTeamTab.tsx:342]:
```suggestion
  const updatedHosts = (changedHosts: Host[]) => {
    const existingHosts = getValues("hosts");
    return changedHosts.map((newValue) => {
      const existingHost = existingHosts.find((host: Host) => host.userId === newValue.userId);
      return existingHost
        ? {
            ...newValue,
            scheduleId: existingHost.scheduleId,
          }
        : newValue;
    });
  };

```

Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventTeamTab.tsx]: added comment and updated

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx:393]: though key prop was added , warning was occuring due to empty parent at 398, removed it

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (Amit91848) [inline apps/web/components/eventtype/EventAvailabilityTab.tsx:366]: NIT: 
```suggestion
        {isPending ? (
          <Spinner className="mt-2 h-6 w-6" />
        ) : (
          <Controller
            name={`hosts.${index}.scheduleId`}
            render={({ field }) => {
              return (
                <Select
                  placeholder={t("select")}
                  options={options}
                  isSearchable={false}
                  onChange={(selected) => {
                    field.onChange(selected?.value || null);
                  }}
                  className="block w-full min-w-0 flex-1 rounded-sm text-sm"
                  value={value as AvailabilityOption}
                  components={{ Option, SingleValue }}
                  isMulti={false}
                  isDisabled={isPending}
                />
              );
            }}
          />
        )}
```

**Review by Amit91848 — DISMISSED** [review]
You can use defualtScheduleId in teamMembers here, in most cases it would be there from my understanding (when user has completed the onboarding), otherwise we are falling back to default schedule anyways.

https://github.com/calcom/cal.com/blob/c68664d44284912c8d5c0327efa953712af7e6b4/apps/web/components/eventtype/EventTeamTab.tsx#L422-L424

Otherwise changes look good

**Comment by Amit91848** [comment]
@ciaranha can you confirm the designs. Also it works for both Round Robin and Collective, is this something we want or should this be restricted to Round Robin only ?

File: apps/web/components/eventtype/EventAvailabilityTab.tsx
Comment (vijayraghav-io) [reply apps/web/components/eventtype/EventAvailabilityTab.tsx]: done

**Comment by vijayraghav-io** [comment]
> @ciaranha can you confirm the designs. Also it works for both Round Robin and Collective, is this something we want or should this be restricted to Round Robin only ?

@ciaranha can you please confirm the designs / share your comments.

**Comment by ciaranha** [comment]
> @ciaranha can you please confirm the designs / share your comments.

@vijayraghav-io thanks, this is great for now!

@itsberkaya I would say you could do a quick design pass at this tab and establish better spacing and clean it up a bit?
CleanShot 2024-07-30 at 07 15 33@2x

**Comment by vijayraghav-io** [comment]
> > @ciaranha can you please confirm the designs / share your comments.
> 
> @vijayraghav-io thanks, this is great start on the UX.
> 
> @itsberkaya Could do a design pass at this tab and clean it up a bit? ![CleanShot 2024-07-30 at 07 15 33@2x]

Thankyou @ciaranha , for your feedback.

**Comment by vijayraghav-io** [comment]
Thankyou @ciaranha!

@itsberkaya can you please do the needful, here. Thankyou!

**Comment by vijayraghav-io** [comment]
Hi @Amit91848 / @ciaranha ,
Can you please approve from your ends, as this PR is functional (done few rounds of self-review, functional testing and also tested by Amit during reviews ) as per requirement and design also seems to be accepted except few cleanups.
Thank you!
cc : @keithwillcode

**Comment by Amit91848** [comment]
Hey @vijayraghav-io, sorry for the delay but we are having internal discussion and finalizing the specs for the feature. There are also some design concerns being looked into by itsberkaya. 

Also since this is a high risk feature, it will have to go through another round of reviews from some senior team member. This might take some while, please bear with us. We are trying to do it as quickly as possible.

**Comment by vijayraghav-io** [comment]
Thankyou! @Amit91848 for the update.

**Comment by vijayraghav-io** [comment]
@ciaranha 
Updated for changes due to weights functionality.
Also updated the spacing in Availability tab, hope this is fine.
https://www.loom.com/share/b194ce55a54649edad65eb10587994c6?sid=3faf6346-e165-4e4d-adf4-a4a9f0193bb7

File: apps/web/public/static/locales/en/common.json
Comment (CarinaWolli) [inline apps/web/public/static/locales/en/common.json:2564]:
```suggestion
  "choose_hosts_schedule" : "Choose host schedules",
```
I think that's better

File: apps/web/public/static/locales/en/common.json
Comment (CarinaWolli) [inline apps/web/public/static/locales/en/common.json:2565]:
```suggestion
  "hosts_schedule_description": "You can select a different schedule for each host. Each host's default schedule is chosen by default.",
```

File: packages/prisma/schema.prisma
Comment (CarinaWolli) [inline packages/prisma/schema.prisma:620]:
```suggestion
  host         Host[]
```

**Review by CarinaWolli — COMMENTED** [review]
@itsberkaya do we already have designs for that? 

If designs are not yet ready I think it can be merged like that and we can do a follow-up with UI improvements

**Comment by CarinaWolli** [comment]
@emrysal could you maybe review as well (high risk PR)

File: packages/prisma/schema.prisma
Comment (vijayraghav-io) [reply packages/prisma/schema.prisma]: Thankyou! @CarinaWolli for the review, i have updated suggested changes in single commit

File: apps/web/public/static/locales/en/common.json
Comment (vijayraghav-io) [reply apps/web/public/static/locales/en/common.json]: updated

File: apps/web/public/static/locales/en/common.json
Comment (vijayraghav-io) [reply apps/web/public/static/locales/en/common.json]: updated

**Comment by olivierlambert** [comment]
Gentle reminder that this PR is a pretty cool one to be merged, fixing some issues ;)

**Comment by vijayraghav-io** [comment]
@Amit91848 , can you please take this PR forward 🙏

**Comment by Amit91848** [comment]
Hey @vijayraghav-io sorry for the delay. Code looks good as far as I am concerned, but will take a look again tomorrow. But this being high risk, will have to wait for someone from foundation team to review and approve it. We have bumped up the priority but they are already swamped so it can take a while.

**Comment by vijayraghav-io** [comment]
> Hey @vijayraghav-io sorry for the delay. Code looks good as far as I am concerned, but will take a look again tomorrow. This pr being high risk, will have to wait for someone from foundation team to review and approve it. We have bumped up the priority but they are already swamped so it can take some time.

Thanks @Amit91848 for update, hoping to happen soon 🙏

File: packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts
Comment (emrysal) [inline packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:42]: Is the best thing to do here unauthorized? It's a situation that should never happen (but could happen) - but not sure that UNAUTHORIZED is the best approach.

Comment (vijayraghav-io) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:42]: Thankyou! @emrysal for starting to review 🙏 

yeah may be is "NOT_FOUND" more apt here? Please let me know if this is ok.
Assuming that there will be at least one schedule for a given user, this may never happen, may be i added in case to handle the error if for some reason Prisma is not able to fetch the data.

Comment (vijayraghav-io) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:42]: it's a error message for very rarely occurring scenario we can update.

Comment (emrysal) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:42]: NOT_FOUND is good, also if you can add a logger to log this? We recently increased our log verbosity settings so it makes sense to log it using the logger.

Comment (vijayraghav-io) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:46]: Thankyou! @emrysal  for confirming, 
updated error code to NOT_FOUND, also added log.

Comment (vijayraghav-io) [reply packages/trpc/server/routers/viewer/availability/schedule/getAllSchedulesByUserId.handler.ts:46]: @emrysal, can you please let know if you have any other comments or it is ok.

**Comment by anikdhabal** [comment]
Hey @vijayraghav-io could you pls fix the conflicts?

**Comment by vijayraghav-io** [comment]
> Hey @vijayraghav-io could you pls fix the conflicts?

Sure, will update.

**Comment by vijayraghav-io** [comment]
@anikdhabal , resolved merge conflicts.

**Comment by emrysal** [comment]
/tip 25 for patience whilst we get through our PR backlog

**Review by emrysal — APPROVED** [review]
Tested this thoroughly on local, all tests passing and happy to merge. 👏

**Comment by vijayraghav-io** [comment]
> Tested this thoroughly on local, all tests passing and happy to merge. 👏

Thankyou! @emrysal  🙏

**Comment by emrysal** [comment]
/tip 25 @vijayraghav-io for patience whilst we get through our PR backlog

**Comment by emrysal** [comment]
/tip $25 @vijayraghav-io for patience whilst we get through our PR backlog

**Pull request merged**
